File: frontend/src/api/e2ee.ts
import apiClient from './client'

// ─── Types ─────────────────────────────────────────────────────────────────

export type QRSessionStatus = 'pending' | 'scanned' | 'approved' | 'expired' | 'used'

export interface QRChallengeResponse {
  session_id: string
  device_ephemeral_pk: string
  expires_at: string
}

export interface QRStatusResponse {
  session_id: string
  status: QRSessionStatus
  // Only present when status === "approved"
  access_token?: string
  refresh_token?: string
  encrypted_private_key?: string
  encryption_nonce?: string
  approver_e2ee_public_key?: string
}

export interface E2EEPublicKeyRead {
  user_id: string
  public_key: string
  fingerprint: string
  updated_at: string
}

// ─── QR session ────────────────────────────────────────────────────────────

/**
 * Create a new QR login session on the server.
 * Called by the NEW (untrusted) device.
 *
 * @param deviceEphemeralPk  Base64 SPKI of the new device's ephemeral ECDH public key
 */
export async function createQRChallenge(deviceEphemeralPk: string): Promise<QRChallengeResponse> {
  const { data } = await apiClient.post('/auth/qr/challenge', {
    device_ephemeral_pk: deviceEphemeralPk,
  })
  return data
}

/**
 * Poll the status of a QR session.
 * Called by the NEW device every ~2 s until status is "approved" or "expired".
 */
export async function pollQRStatus(sessionId: string): Promise<QRStatusResponse> {
  const { data } = await apiClient.get(`/auth/qr/${sessionId}/status`)
  return data
}

/**
 * Approve a QR login session from the trusted (phone) device.
 *
 * @param sessionId             The session ID from the QR payload
 * @param encryptedPrivateKey   AES-GCM ciphertext of the user's E2EE private key
 * @param encryptionNonce       AES-GCM IV (nonce)
 * @param approverE2eePublicKey Approver's E2EE ECDH public key (SPKI base64)
 */
export async function approveQRSession(
  sessionId: string,
  encryptedPrivateKey: string,
  encryptionNonce: string,
  approverE2eePublicKey: string,
): Promise<QRStatusResponse> {
  const { data } = await apiClient.post(`/auth/qr/${sessionId}/approve`, {
    encrypted_private_key: encryptedPrivateKey,
    encryption_nonce: encryptionNonce,
    approver_e2ee_public_key: approverE2eePublicKey,
  })
  return data
}

// ─── E2EE public key management ────────────────────────────────────────────

/** Upload/rotate your own E2EE public key. */
export async function publishPublicKey(publicKey: string, fingerprint: string): Promise<E2EEPublicKeyRead> {
  const { data } = await apiClient.put('/me/e2ee-public-key', { public_key: publicKey, fingerprint })
  return data
}

/** Fetch another user's E2EE public key (to encrypt a DM for them). */
export async function getUserPublicKey(userId: string): Promise<E2EEPublicKeyRead | null> {
  try {
    const { data } = await apiClient.get(`/users/${userId}/e2ee-public-key`)
    return data
  } catch (err: any) {
    if (err?.response?.status === 404) return null
    throw err
  }
}

File: backend/models/e2ee.py
"""Models for E2EE key management and QR-code-based device login."""
import uuid
import enum
from datetime import datetime, timezone

from sqlalchemy import String, Text, DateTime, ForeignKey, Enum, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship

from models.base import Base


class QRSessionStatus(str, enum.Enum):
    pending = "pending"    # waiting for trusted device to scan
    scanned = "scanned"    # trusted device scanned, waiting for user confirmation
    approved = "approved"  # user confirmed — new device may log in + extract key
    expired = "expired"    # TTL exceeded before approval
    used = "used"          # new device already consumed the session


class QRSession(Base):
    """Represents a pending QR-code login session.

    Flow:
      1. New device POST /auth/qr/challenge  → creates a QRSession (status=pending)
         and includes the device's ephemeral ECDH public key.
      2. QR code displayed on new device encodes {session_id, device_ephemeral_pk}.
      3. Trusted device (phone) POST /auth/qr/{id}/approve:
         - authenticates via its own access token
         - encrypts the user's E2EE private key under the new device's ephemeral key
         - server mints a fresh access+refresh token pair for the new session
         - session status → approved; tokens + encrypted key stored on session row
      4. New device polls GET /auth/qr/{id}/status until approved, then:
         - reads encrypted_private_key + phone_e2ee_public_key from response
         - derives shared ECDH secret, decrypts private key
         - uses access_token + refresh_token to authenticate
    """

    __tablename__ = "qr_sessions"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    status: Mapped[QRSessionStatus] = mapped_column(
        String(20),
        default=QRSessionStatus.pending,
        nullable=False,
    )

    # The ECDH public key of the new device (ephemeral, base64-encoded DER/SPKI)
    device_ephemeral_pk: Mapped[str] = mapped_column(Text, nullable=False)

    # Filled in by the approving (trusted) device
    approver_user_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=True
    )
    # AES-GCM encrypted E2EE private key (base64 SPKI/PKCS8)
    encrypted_private_key: Mapped[str | None] = mapped_column(Text, nullable=True)
    # AES-GCM nonce for the encrypted private key (base64)
    encryption_nonce: Mapped[str | None] = mapped_column(String(64), nullable=True)
    # The approver's E2EE ECDH public key (needed by new device to derive shared secret)
    approver_e2ee_public_key: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Token pair minted for the new session on approval
    access_token: Mapped[str | None] = mapped_column(Text, nullable=True)
    refresh_token: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    approver: Mapped["User | None"] = relationship("User", foreign_keys=[approver_user_id])


class UserE2EEKey(Base):
    """Stores a user's E2EE ECDH public key, published so other users can encrypt messages for them."""

    __tablename__ = "user_e2ee_keys"

    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"),
        primary_key=True,
    )
    # Base64-encoded SPKI public key bytes (ECDH P-256)
    public_key: Mapped[str] = mapped_column(Text, nullable=False)
    # Human-readable fingerprint (first 16 hex chars of SHA-256 of the raw key bytes)
    fingerprint: Mapped[str] = mapped_column(String(64), nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    user: Mapped["User"] = relationship("User", foreign_keys=[user_id])
